File: Toggly.FeatureManagement.Java/toggly-core/src/main/java/io/toggly/core/SdkIdentity.java
package io.toggly.core;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * SDK identity constants for HTTP User-Agent and WebSocket query parameters.
 */
public final class SdkIdentity {

    public static final String SDK_ID = "java";
    public static final String SDK_VERSION = "1.0.1";

    private SdkIdentity() {
    }

    public static String userAgent() {
        return "toggly-" + SDK_ID + "/" + SDK_VERSION;
    }

    public static String appendSdkQueryParams(String wsUrl, String cachedRevision) {
        Map<String, String> params = new LinkedHashMap<>();
        if (cachedRevision != null && !cachedRevision.isEmpty()) {
            params.put("rev", cachedRevision);
        }
        params.put("sdk", SDK_ID);
        params.put("sdkVersion", SDK_VERSION);

        String query = params.entrySet().stream()
                .map(entry -> URLEncoder.encode(entry.getKey(), StandardCharsets.UTF_8)
                        + "="
                        + URLEncoder.encode(entry.getValue(), StandardCharsets.UTF_8))
                .collect(Collectors.joining("&"));

        return wsUrl + "?" + query;
    }
}

File: Toggly.FeatureManagement.Android/toggly-android-core/src/main/kotlin/io/toggly/core/SdkIdentity.kt
package io.toggly.core

import java.net.URLEncoder

object SdkIdentity {
    const val SDK_ID = "android"
    const val SDK_VERSION = "1.1.0"

    fun userAgent(): String = "toggly-$SDK_ID/$SDK_VERSION"

    fun appendSdkQueryParams(wsUrl: String, cachedRevision: String?): String {
        val params = linkedMapOf<String, String>()
        if (!cachedRevision.isNullOrEmpty()) {
            params["rev"] = cachedRevision
        }
        params["sdk"] = SDK_ID
        params["sdkVersion"] = SDK_VERSION

        val query = params.entries.joinToString("&") { (key, value) ->
            "${URLEncoder.encode(key, "UTF-8")}=${URLEncoder.encode(value, "UTF-8")}"
        }

        return "$wsUrl?$query"
    }
}
